CMake rework - modernization and simplification

kreczko opened the pull request:
@tunnell @jpbrodsky @jecutter 
This arose from a private discussion with @mszydagis on LZ gitlab.

## Summary
While looking at the 2.1.0beta I realised that there is a difference in how headers are stored in the repo and how they are installed. The repository structure is closer to current standards that scope headers (e.g. `#include "NEST/NEST.hh"` vs `#include "NEST.hh"`.

I took this opportunity to fix a few things in the current CMake as well as to modernize (e.g. `NEST::Core` target) and simplify (current CMake is a mix of old and new language styles).

With a simple
```cmake
target_link_libraries(<your target> PUBLIC NEST::Core)
```
all necessary include paths are imported and allow for both nested and direct header inclusions:
```c++
#include "NEST.hh"
#include "NEST/NEST.hh"

#include "VDetector.hh"
#include "Detectors/VDetector.hh"

#include "NEST/G4/NESTStackingAction.hh"
#include "NESTStackingAction.hh"
```

Ideally, `Detectors` would also go under `NEST/`, but that would mismatch the repository structure.
Full install structure is shown below.

The old CMake produced both STATIC (only used for local executables) and SHARED versions of the core library, the current produces STATIC by default and SHARED with [BUILD_SHARED_LIBS=ON](https://cmake.org/cmake/help/v3.0/variable/BUILD_SHARED_LIBS.html). If the old behaviour is required, I can add an object library as an intermediate step.

Feedback is welcome.

## Current 2.1.0beta install structure (CentOS7)
```bash
 tree -L 2 install/
install/
├── NESTConfig-noconfig.cmake
├── NESTConfig.cmake
├── NESTConfigVersion.cmake
├── bareNEST
├── include
│   ├── DetectorExample_XENON10.hh
│   ├── LUX_Run03.hh
│   ├── NEST.hh
│   ├── RandomGen.hh
│   ├── TestSpectra.hh
│   ├── VDetector.hh
│   ├── Xed_CWRU_Dahl.hh
│   ├── Z3_FSR_HA.hh
│   ├── analysis.hh
│   ├── testNEST.hh
│   └── xoroshiro.hh
├── lib
│   └── libNEST.so
└── testNEST
```

## Install structure with this MR (G4=ON, BUILD_ROOT=ON)

```bash
```shell
 tree -L 4 install/
install/
├── bin
│   ├── bareNEST
│   ├── rootNEST
│   └── testNEST
├── include
│   ├── Detectors
│   │   ├── DetectorExample_XENON10.hh
│   │   ├── LUX_Run03.hh
│   │   ├── VDetector.hh
│   │   ├── Xed_CWRU_Dahl.hh
│   │   └── Z3_FSR_HA.hh
│   └── NEST
│       ├── G4
│       │   ├── NESTProc.hh
│       │   └── NESTStackingAction.hh
│       ├── NEST.hh
│       ├── RandomGen.hh
│       ├── TestSpectra.hh
│       ├── analysis.hh
│       ├── testNEST.hh
│       └── xoroshiro.hh
└── lib64
    ├── cmake
    │   └── NEST
    │       ├── NESTConfig.cmake
    │       ├── NESTConfigVersion.cmake
    │       ├── NESTTargets-noconfig.cmake
    │       └── NESTTargets.cmake
    ├── libCore.a
    └── libNESTG4.a
```

jpbrodsky commented on G4integration/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+)
+add_library(NEST::NESTG4 ALIAS NESTG4)
+
+target_compile_features(NESTG4 PUBLIC cxx_std_17)
```
Again, does this need to be 17?

kreczko replied:
Removed

jpbrodsky commented on CMakeLists.txt:
```diff
@@ -1,119 +1,116 @@
+  $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include/Detectors>
+)
+
+target_compile_features(Core PUBLIC cxx_std_17)
```
(Notably, ROOT is often compiled with 14 or even 11 with non-GCC compilers, due to weirdness in how ROOT handles its compiler feature checks)

jpbrodsky replied:
So I wouldn't force this into 17 if it's not necessary.

kreczko replied:
it is not necessary - will remove.

mszydagis commented:
One more thing I need your help with on this, @kreczko and @jpbrodsky  I have added a new executable (like rootNEST) in the examples sub-folder. But I have been manually, for weeks now, compiling it awkwardly just separately on my own. It is called PulseShape.cpp  I don't know enough about making makefiles (pun intended haha!) Could one of you please add this new pulse shape code to the cmake file, alongside rootNEST (and testNEST and bareNEST etc.) in same way? Thank you!! (And then I can tag the release tomorrow hopefully, woo-hoo!) This will also add the advantage of being able to call headers and such from everywhere. Thanks again

kreczko commented:
@mszydagis 
This is now done in d347e57.

For future reference, this is the pattern for the example directory:
In `examples/CMakeLists.txt` add

```cmake
add_executable(<binary name> <sources>)
target_link_libraries(<binary name>PUBLIC NEST::Core)
list(APPEND EXAMPLES <binary name>)
```

BTW: PulseShape.cpp was not compiling (missing `#include <algorithm>` for `std::sort`) - I've rebased onto master and fixed the missing header.

mszydagis commented:
thank you for this. And good catch on missing header: for some reason worked on my laptop.
